lib/decompress_impl.ml

```diff
@@ -1538,6 +1555,350 @@ module Zlib_deflate = struct
+    | None -> None
+
+  let digest_crc16_string str crc16 =
+    match crc16 with
+    | Some crc16 ->
+        Some (Checkseum.Crc32.digest_string str 0 (String.length str) crc16)
+    | None -> None
+
+  module KHeader = struct
+    let ctor k = Header k
+
+    let put_short_lsb short k src dst t =
+      let crc16 = digest_crc16_byte (short land 0xFF) t.crc16 in
+      let crc16 = digest_crc16_byte ((short lsr 8) land 0xFF) crc16 in
+      put_short_lsb ~ctor short k src dst {t with crc16}
+
+    let put_byte byte k src dst t =
+      let crc16 = digest_crc16_byte byte t.crc16 in
+      put_byte ~ctor byte k src dst {t with crc16}
+
+    let put_string str k src dst t =
+      let crc16 = digest_crc16_string str t.crc16 in
+      put_string ~ctor str k src dst {t with crc16}
+  end
+
+  module KSize = struct
+    let ctor k = Size k
+    let align k src dst t = align ~ctor k src dst t
+    let put_short_lsb short k src dst t = put_short_lsb ~ctor short k src dst t
+  end
+
+  module KCrc32 = struct
+    let ctor k = Crc32 k
+    let align k src dst t = align ~ctor k src dst t
+    let put_short_lsb short k src dst t = put_short_lsb ~ctor short k src dst t
+  end
+
+  let size src dst t =
+    let size = RFC1951_deflate.read_in t.d in
+    let k _src _dst t = ok t in
+    ( KSize.align
+    @@ KSize.put_short_lsb Int32.(to_int (Int32.logand size 0xFFFFl))
+    @@ KSize.put_short_lsb
+         Int32.(to_int (Int32.logand (Int32.shift_right size 16) 0xFFFFl))
+         k )
+      src dst t
+
+  let crc32 src dst t =
+    let crc = t.d.RFC1951_deflate.crc in
+    let k _src _dst t = Cont {t with z= Size size} in
+    ( KCrc32.align
+    @@ KCrc32.put_short_lsb Optint.(to_int Infix.(crc && of_int32 0xFFFFl))
+    @@ KCrc32.put_short_lsb
+         Optint.(to_int Infix.(crc >> 16 && of_int32 0xFFFFl))
+         k )
+      src dst t
+
+  let deflate src dst t =
+    match RFC1951_deflate.eval0 src dst t.d with
+    | RFC1951_deflate.Cont d -> Cont {t with d}
+    | RFC1951_deflate.Wait d -> Wait {t with d}
+    | RFC1951_deflate.Flush d -> Flush {t with d}
+    | RFC1951_deflate.Ok d -> Cont {t with z= Crc32 crc32; d}
+    | RFC1951_deflate.Error (d, exn) -> error {t with d} (RFC1951 exn)
+
+  let nop k src dst t = k src dst t
+
+  let fextra extra k src dst t =
+    let extra_l = String.length extra in
+    ( KHeader.put_short_lsb extra_l
+    @@ KHeader.put_string extra
+    @@ fun src dst t -> k src dst t )
+      src dst t
+
+  let fname name k src dst t =
+    ( KHeader.put_string name
+    @@ KHeader.put_byte 0
+    @@ fun src dst t -> k src dst t )
+      src dst t
+
+  let fcomment comment k src dst t =
+    ( KHeader.put_string comment
+    @@ KHeader.put_byte 0
+    @@ fun src dst t -> k src dst t )
+      src dst t
+
+  let fcrc16 crc16 k src dst t =
+    ( KHeader.put_short_lsb (Optint.to_int crc16)
+    @@ fun src dst t -> k src dst t )
+      src dst t
+
+  let get_option k = function Some opt -> k opt | None -> nop
```

Comment:
First, it seems than `get_option` is only used in `header` function. It better to put it inside than to let it on the global scope. Then, `get_option` is not a good name, may be we should rename it to something more explicit. Finally, this kind of function is close to what an `Option` module should provide.

Because we already talk about some `Option` functions, it should be more interesting to provide a module like that instead to have some random function about `option` type in the code.